src/vos/ilog.c

```diff
@@ -0,0 +1,1474 @@
+					     DAOS_INTENT_UPDATE);
+
+	if (visibility == ILOG_REMOVED || ilog_empty(root)) {
+		if (opc != ILOG_OP_UPDATE) {
+			D_DEBUG(DB_IO, "ilog entry "DF_U64" not found\n",
+				id_in->id_epoch);
+			goto done;
+		}
+
+		if (visibility == ILOG_REMOVED) {
+			D_DEBUG(DB_IO, "Removing aborted "DF_U64
+				" from ilog root\n", root->lr_id.id_epoch);
+			old_id = root->lr_id;
+			tmp.lr_magic = ILOG_MAGIC;
+			if (opc != ILOG_OP_UPDATE) {
+				rc = ilog_ptr_set(lctx, root, &tmp);
+				goto remove;
+			}
+		}
+
+		D_DEBUG(DB_IO, "Inserting "DF_U64" at ilog root\n",
+			id_in->id_epoch);
+		tmp.lr_magic = ILOG_MAGIC;
+		tmp.lr_id.id_epoch = id_in->id_epoch;
+		tmp.lr_punch = punch;
+		rc = ilog_ptr_set(lctx, root, &tmp);
+		if (rc != 0)
+			goto done;
+		rc = ilog_register_tx(lctx, &root->lr_id);
+		if (rc != 0)
+			goto done;
+remove:
+		/** Remove the aborted tx_id record */
+		rc = ilog_deregister_tx(lctx, &old_id);
+	} else if (root->lr_tree.it_embedded) {
+		bool	is_equal;
+
+		rc = update_inplace(lctx, &root->lr_id, &root->lr_punch,
+				    id_in, opc, punch, &is_equal);
+		if (rc != 0)
+			goto done;
+
+		if (is_equal) {
+			if (opc == ILOG_OP_ABORT) {
+				D_DEBUG(DB_IO, "Removing "DF_U64
+					" from ilog root\n", id_in->id_epoch);
+				tmp.lr_magic = ILOG_MAGIC;
+				rc = ilog_ptr_set(lctx, root, &tmp);
+			}
+			goto done;
+		}
+
+		if (opc != ILOG_OP_UPDATE) {
+			D_DEBUG(DB_IO, "Entry "DF_U64" not found in ilog\n",
+				id_in->id_epoch);
+			goto done;
+		}
+
+		if (!punch && !root->lr_punch &&
+		    id_in->id_epoch > root->lr_id.id_epoch &&
+		    visibility == ILOG_VISIBLE) {
+			D_DEBUG(DB_IO, "No update needed\n");
+			goto done;
+		}
+		/* Either this entry is earlier or prior entry is uncommitted
+		 * or either entry is a punch
+		 */
+		rc = ilog_root_migrate(lctx, id_in->id_epoch, punch);
+	} else {
+		/** Ok, we have a tree.  Do the operation in the tree */
+		rc = ilog_tree_modify(lctx, id_in, punch, epr, opc);
+	}
+done:
+	rc = ilog_tx_end(lctx, rc);
+	D_DEBUG(DB_IO, "%s in incarnation log "DF_U64" status: rc=%s\n",
+		opc_str[opc], id_in->id_epoch, d_errstr(rc));
+	return rc;
+}
+
+int
+ilog_update(daos_handle_t loh, daos_epoch_t epoch, bool punch)
+{
+	daos_epoch_range_t	 range = {0, DAOS_EPOCH_MAX};
+	struct ilog_id		 id = {0, epoch};
+
+	return ilog_modify(loh, &id, punch, &range, ILOG_OP_UPDATE);
+
+}
+
+/** Makes a specific update to the incarnation log permanent and
+ *  removes redundant entries
+ */
+int
+ilog_persist(daos_handle_t loh, const struct ilog_id *id)
+{
+	daos_epoch_range_t	 range = {0, DAOS_EPOCH_MAX};
+
+	return ilog_modify(loh, id, false, &range, ILOG_OP_PERSIST);
+}
+
+/** Removes a specific entry from the incarnation log if it exists */
+int
+ilog_abort(daos_handle_t loh, const struct ilog_id *id)
+{
+	daos_epoch_range_t	 range = {0, DAOS_EPOCH_MAX};
+
+	return ilog_modify(loh, id, false, &range, ILOG_OP_ABORT);
+}
+
+int
+ilog_aggregate(daos_handle_t loh, const daos_epoch_range_t *epr)
```

Comment:
So this will be called by VOS aggregation in follow-on patch?

Reply:
yes.   And probably modified a bit :-)